coffee_boat/captain.py

```diff
@@ -0,0 +1,166 @@
+    name: {0}
+    dependencies:
+    - python=={1}
+    - anaconda
+    - pip
+    - pip:
+    """).format(self.env_name, self.python_version)
+    package_spec = "{0}{1}".format(base_package_spec, pip_packages)
+    package_spec_file = tempfile.NamedTemporaryFile(dir=self.working_dir, delete=handle_del)
+    package_spec_path = package_spec_file.name
+    print("Writing package spec to {0}.".format(package_spec_path))
+    package_spec_file.write(package_spec)
+    package_spec_file.flush()
+    # Create the conda env
+    conda_prefix = os.path.join(self.working_dir, self.env_name)
+    print("Creating conda env")
+    if os.path.exists(conda_prefix):
+      print("Cleaining up old prefix {0}".format(conda_prefix))
+      subprocess.check_call(["rm", "-rf", conda_prefix])
+    subprocess.check_call([self.conda, "env", "create", "-f", package_spec_path,
+                           "--prefix", conda_prefix], stdout=DEVNULL)
+    # Package it for distro
+    zip_target = os.path.join(self.working_dir, "coffee_boat.zip")
+    print("Packaging conda env")
+    subprocess.check_call(["zip", zip_target, "-r", conda_prefix], stdout=DEVNULL)
+    relative_python_path = "." + conda_prefix + "/bin/python"
+
+    # Make a self extractor script
+    runner_script = inspect.cleandoc("""#!/bin/bash
+    if [ -f coffee_boat.zip ];
+    then
+      unzip coffee_boat.zip &>/dev/null && rm coffee_boat.zip &> /dev/null
+    fi
+    {0} "$@" """.format(relative_python_path))
+    runner_script_path = os.path.join(self.working_dir, "coffee_boat_runner.sh")
+    with open(runner_script_path, 'w') as f:
+      f.write(runner_script)
+    subprocess.check_call(["chmod", "a+x", runner_script_path])
+
+    # Screw around with enviroment variables so that the env gets distributed.
+    old_args = os.environ.get("PYSPARK_SUBMIT_ARGS", "pyspark-shell")
+    new_args = "--py-files {0},{1} {2}".format(zip_target, runner_script_path, old_args)
+    os.environ["PYSPARK_SUBMIT_ARGS"] = new_args
+    if "PYSPARK_GATEWAY_PORT" in os.environ:
+      print("Hey the Java process is already running, this might not work.")
+    os.environ["PYSPARK_PYTHON"] = "./coffee_boat_runner.sh"
+
+
+
+  def _raise_if_running(self):
+    """Raise an exception if Spark is already running because its too late to add or package
+    dependencies at this point."""
+    if pyspark.SparkContext._active_spark_context is not None:
+      raise Exception("Spark context is already running. "
+                      "All coffee boat activites must occure before launching your SparkContext."
+                      "You can stop your current SparkContext (with sc.stop() or session.stop()) "
+                      "add more dependencies and re-start your SparkContext if needed.")
+
+
+  def _setup_or_find_conda(self):
+    # Check if we need to setup conda or return if we already have one
+    rc = subprocess.call(['which', 'conda'])
+    if rc == 0:
+      self.conda = "conda"
```

Comment:
Does it matter which version of `conda` is available? I'm thinking at the point at which someone installs this across the cluster in advance, they may want to specify which version is installed across the cluster or used locally.